### PR TITLE
⌨️ refactor UseFormMethods to arrow functions

### LINKED ISSUE #2887: ESlint TypeScript error `unbound-method`
**Describe the bug**
Deconstructing `register` from `useForm` triggers `@typescript-eslint` error [`unbound-method`](https://github.com/typescript-eslint/typescript-eslint/blob/v4.1.0/packages/eslint-plugin/docs/rules/unbound-method.md)

**To Reproduce**
Use ESlint with a rule mentioned above.

**Codesandbox link (Required)**
CodeSandbox does not support ESlint (https://github.com/codesandbox/codesandbox-client/issues/1933)

**Expected behavior**
`register` is not a method that depends on `this` context, so should not trigger a warning. **`handleSubmit` doesn't**.

**Screenshots**
Screenshot from 2020-09-11 15-40-37

**Desktop (please complete the following information):**

- OS: btw, I use Arch (not really, Manjaro)
- Browser Vivaldi
- Version up to date
:P

**Additional context**
Similar problem with Formik: https://github.com/formium/formik/issues/2589

*PR welcome? I can contribute*

### PR DESCRIPTION
**Pull request by Tymek**
resolves #2887

### CODE REVIEW

**Comment by Tymek** [comment]
I got the main issue down, but I don't understand why TypeScript complains about generics in multiple invocation signatures while it was fine using overloaded functions :|

**Comment by bluebill1049** [comment]
Thanks @Tymek I will take a look at it.

File: src/useForm.ts [hunk 1]
```diff
@@ -514,7 +514,7 @@ export function useForm<
     TFieldValue extends TFieldValues[TFieldName]
   >(
     name: TFieldName,
-    value: NonUndefined<TFieldValue> extends NestedValue<infer U>
+    value?: NonUndefined<TFieldValue> extends NestedValue<infer U>
```
Comment (bluebill1049) [inline src/useForm.ts:517]:
```suggestion
    value: NonUndefined<TFieldValue> extends NestedValue<infer U>
```

**Comment by bluebill1049** [comment]
actually @kotarella1110 did you had a similar issue before? I remember from one of the PR. it was to do with the type order.

**Comment by kotarella1110** [comment]
@bluebill1049 yes. I had a similar problem before.
The solution was to change the order of the methods in the `Control` type.
The cause of this is not well understood...
I tested this and this passes... strange...

**Comment by bluebill1049** [comment]
Thanks @kotarella1110 i will play with it a bit more.

**Comment by bluebill1049** [comment]
I am having issue to get this one over the line as well, attached with help wanted tag, hopefully someone can figure out the reason and i will take a look at it again as well.

**Comment by bluebill1049** [comment]
It's a super crap this one, as unit tests pass locally.

**Comment by bluebill1049** [comment]
my 🧠 is fried... @kotarella1110 if you some free time, please help to take a look at this...

**Comment by Tymek** [comment]
I did some more debugging and.. this is madness

maybe we can create matching Control<Type> from `name`, instead of trying to infer it?
what this type check does, is checking if right `name` exists on `Control`
inverting this will check _if we can use passed `Control` with this `name`_

**Comment by Tymek** [comment]
I have no idea why tests passed locally, but I had the same problem. I've run it in docker with `docker run -it -v ${PWD}:/srv -w /srv circleci/node:lts bash` instead of loosing my sanity asking what are the differences between `yarn test` and `npm test`.

**Comment by bluebill1049** [comment]
@Tymek I got my 🧠 fried by this issue... I couldn't find a clue why tests failed everyone and then, while e2e constantly(build) failed. I am not even sure it's our code or ts...

**Comment by JeromeDeLeon** [comment]
This is the only time we could probably say that `It works on my local`. I have tried them out myself but no luck.

**Comment by bluebill1049** [comment]
Thanks for trying to solve this "hard" problem @Tymek 🙏 ❤️

**Comment by Tymek** [comment]
Your're welcome :wink: I'm glad that I can help

CI tests finally passed :tada:

**Comment by bluebill1049** [comment]
oh wow! thank you! the power of OSS.

File: src/controller.tsx [hunk 2]
```diff
@@ -7,8 +7,7 @@ import skipValidation from './logic/skipValidation';
 import isNameInFieldArray from './logic/isNameInFieldArray';
 import { useFormContext } from './useFormContext';
 import { VALUE } from './constants';
-import { Control } from './types';
-import { ControllerProps } from './types';
+import { ControllerProps, FieldValues } from './types';
```
Comment (bluebill1049) [inline src/controller.tsx:10]: nice~

File: src/controller.tsx [hunk 3]
```diff
@@ -27,8 +26,8 @@ const Controller = <
   control,
   onFocus,
   ...rest
-}: ControllerProps<TAs, TControl>) => {
-  const methods = useFormContext();
+}: ControllerProps<TAs, TFieldValues>) => {
+  const methods = useFormContext<TFieldValues>();
```
Comment (bluebill1049) [inline src/controller.tsx:30]: 👍

Comment (Tymek) [reply src/controller.tsx:30]: in theory changing this might break someone's build, but it's a niche case

**Review by kotarella1110 — APPROVED** [review]
wow! amazing 🏆
looks so good! 💯 
Thank you very much!!!!!

**Comment by bluebill1049** [comment]
Thank you very much @Tymek 🎖️ and @kotarella1110 for the review. 🙏

**Comment by bluebill1049** [comment]
Looks like this PR caused some issues... I am temporally rever it back 🙏

**Pull request merged**